HARK/ConsumptionSaving/ConsLabeledModel.py

```diff
@@ -0,0 +1,663 @@
+        )
+
+        return a
+
+    def value_transition(self, action=None, state=None, continuation=None, params=None):
+        """
+        value of action given state and continuation
+        """
+        variables = {}  # pytree
+        ps = self.state_transition(state, action, params)
+        variables.update(ps)
+
+        variables["reward"] = self.u(action["cNrm"])
+        variables["v"] = variables["reward"] + params.Discount * continuation(ps)
+        variables["v_inv"] = self.u.inv(variables["v"])
+
+        variables["marginal_reward"] = self.u.der(action["cNrm"])
+        variables["v_der"] = variables["marginal_reward"]
+        variables["v_der_inv"] = action["cNrm"]
+
+        # for estimagic purposes
+        variables["contributions"] = variables["v"]
+        variables["value"] = np.sum(variables["v"])
+
+        return variables
+
+    def continuation_transition(self, post_state=None, value_next=None, params=None):
+        """
+        continuation value function of post_state
+        """
+        variables = {}  # pytree
+        ns = self.post_state_transition(post_state, params)
+        variables.update(ns)
+        variables["v"] = params.PermGroFac ** (1 - params.CRRA) * value_next(ns)
+        variables["v_der"] = (
+            params.Rfree
+            * params.PermGroFac ** (-params.CRRA)
+            * value_next.derivative(ns)
+        )
+
+        variables["v_inv"] = self.u.inv(variables["v"])
+        variables["v_der_inv"] = self.u.inv(variables["v_der"], order=(1, 0))
+
+        # for estimagic purposes
+        variables["contributions"] = variables["v"]
+        variables["value"] = np.sum(variables["v"])
+
+        return variables
+
+    def prepare_to_solve(self):
+        self.create_params_namespace()
+        self.define_boundary_constraint()
+        self.create_post_state()
+
+    def create_continuation_function(self):
+
+        # unpack next period's solution
+        vfunc_next = self.solution_next.value
+
+        v_end = self.continuation_transition(self.post_state, vfunc_next, self.params)
+        # need to drop m because it's next period's m
+        v_end = xr.Dataset(v_end).drop(["mNrm"])
+        borocnst = self.borocnst.drop(["mNrm"]).expand_dims("aNrm")
+        if self.nat_boro_cnst:
+            v_end = xr.merge([borocnst, v_end])
+
+        wfunc = ValueFuncCRRALabeled(v_end, self.params.CRRA)
+
+        return wfunc
+
+    def endogenous_grid_method(self):
+
+        wfunc = self.create_continuation_function()
+
+        # get optimal actions/decisions from egm
+        acted = self.egm_transition(self.post_state, wfunc, self.params)
+        # get state from actions and post_state
+        state = self.reverse_transition(self.post_state, acted, self.params)
+
+        # egm requires swap dimensions; make actions and state functions of state
+        action = xr.Dataset(acted).swap_dims({"aNrm": "mNrm"})
+        state = xr.Dataset(state).swap_dims({"aNrm": "mNrm"})
```

Comment:
You seem to be imputing some semantics on the other order of the dimensions in these DataSets.

What is that semantics?

Reply:
This was a suggestion by @llorracc. An action is a consequence of the state, but given the end of period state, the endogenous grid method tells you how the agent must have acted in order to be optimal. This should be made more clear in documentation.

Reply:
This is cleverly implemented!